src/seraphis_core/legacy_enote_types.cpp

```diff
@@ -51,11 +51,12 @@ const rct::key& onetime_address_ref(const LegacyEnoteVariant &variant)
     struct visitor final : public tools::variant_static_visitor<const rct::key&>
     {
         using variant_static_visitor::operator();  //for blank overload
-        const rct::key& operator()(const LegacyEnoteV1 &enote) const { return enote.onetime_address; }
-        const rct::key& operator()(const LegacyEnoteV2 &enote) const { return enote.onetime_address; }
-        const rct::key& operator()(const LegacyEnoteV3 &enote) const { return enote.onetime_address; }
-        const rct::key& operator()(const LegacyEnoteV4 &enote) const { return enote.onetime_address; }
-        const rct::key& operator()(const LegacyEnoteV5 &enote) const { return enote.onetime_address; }
+        const rct::key& operator()(const LegacyPreRctEnote &enote) const { return enote.onetime_address; }
```

Comment:
Adding a new type here was my suggestion. There needs to be a solution in place in order to handle the following caveat:

All post-RingCT enotes (including post-RingCT coinbase enotes) should use and increment `m_legacy_amount_counts[0]`, whereas all pre-RingCT enotes (including pre-RingCT coinbase enotes) should use and increment `m_legacy_amount_counts[amount]` for the `enote_same_amount_index`. Further, the scanner should capture enough data such that a wallet can construct txs with pre-RingCT enotes in rings composed of pre-RingCT enotes alone (and so it needs to know if an enote is pre-RingCT or post-RingCT).

Because the legacy enote v1 enote type can either be pre-RingCT or a post-RingCT coinbase enote, the v1 type alone doesn't capture enough information such that the scanner can increment the `m_legacy_amount_counts` correctly, nor enough to know to use the enote to construct a pre-RingCT ring with just pre-RingCT enotes (and use the correct index).

Using a distinct enote type for pre-RingCT can enable that.

Looking at the code again, another solution could be a `LegacyEnoteOriginContextV1` for pre-RingCT enotes and `LegacyEnoteOriginContextV2` for post-RingCT. The `LegacyEnoteOriginContextV1` could implement the `enote_same_amount_ledger_index`, and the `LegacyEnoteOriginContextV2` could have a `rct_enote_ledger_index` instead (since coinbase amounts technically aren't the "same amount").

Reply:
Thank you both, already worked on some of the suggestions and will look into adding `LegacyEnoteOriginContextV2` tomorrow. Just thinking, maybe it's a less invasive change to add `rct_enote_ledger_index` to `LegacyEnoteOriginContextV1`!?

Reply:
> Adding a new type here was my suggestion. There needs to be a solution in place in order to handle the following caveat:
> 
> All post-RingCT enotes (including post-RingCT coinbase enotes) should use and increment `m_legacy_amount_counts[0]`, whereas all pre-RingCT enotes (including pre-RingCT coinbase enotes) should use and increment `m_legacy_amount_counts[amount]` for the `enote_same_amount_index`. Further, the scanner should capture enough data such that a wallet can construct txs with pre-RingCT enotes in rings composed of pre-RingCT enotes alone (and so it needs to know if an enote is pre-RingCT or post-RingCT).
> 
> Because the legacy enote v1 enote type can either be pre-RingCT or a post-RingCT coinbase enote, the v1 type alone doesn't capture enough information such that the scanner can increment the `m_legacy_amount_counts` correctly, nor enough to know to use the enote to construct a pre-RingCT ring with just pre-RingCT enotes (and use the correct index).
> 
> Using a distinct enote type for pre-RingCT can enable that.
> 
> Looking at the code again, another solution could be a `LegacyEnoteOriginContextV1` for pre-RingCT enotes and `LegacyEnoteOriginContextV2` for post-RingCT. The `LegacyEnoteOriginContextV1` could implement the `enote_same_amount_ledger_index`, and the `LegacyEnoteOriginContextV2` could have a `rct_enote_ledger_index` instead (since coinbase amounts technically aren't the "same amount").

What can also capture all information needed without a separation of `LegacyEnoteOriginContextV1` and `LegacyEnoteOriginContextV2` is to include the ledger indexing amount.

Reply:
The ledger indexing amount approach still needs a solution for this here: https://github.com/UkoeHB/monero/pull/42#discussion_r1571826028

But generally it does seem it's a cleaner approach comparing the 2 PR's

Reply:
Yes. This is way I want to merge these two PR's: it'll make both simpler. We can just have one `LegacyEnoteOriginContext` with a simpler structure, and CLSAGs are based on actually used indexing.

Reply:
So you're saying merge your PR first, then @SNeedlewoods changes their PR to use the single `LegacyEnoteOriginContext` + `ledger_indexing_amount` and fix (v1 and) v4 enote handling in that spot of the PR?

Reply:
Ah, I just read your comments in no-wallet-left-behind proposing you'll combine the 2 PR's, nice :) thank you